openlibrary/macros/databarWork.html

```diff
@@ -59,7 +59,8 @@ <h3 class="header">
         $if editions_page:
           $ render_times['databarWork: Book provider'] = time()
           $ book_provider = get_book_provider(page)
-          $if book_provider:
+          $ availability = (page.availability or {}) if hasattr(page, 'availability') else {}
+          $if book_provider and (availability.get('is_readable') or availability.get('status') == 'open'):
```

Comment:
```suggestion
          $if book_provider and (availability.get('is_readable') or availability.get('status') == 'open') or book_provider.get_access(page).to_solr_str() == "public":

```
This looks like a good start, but because of the reliance on availability here, it won't work for providers without an `ocaid`, such as Standard Ebooks or OpenStax. It will take more thought, but off the cuff something like the above might get closer.